### Etudiant 1: GIANG Phuong Thu, Cécile (3530406)
### Etudiant 2: KHALFAT Célina (28716860)

# [AMAL] TME 10: Mécanisme d'attention propre

In [1]:
from tp10 import *

INFO:root:Loading embeddings
INFO:root:Get the IMDB dataset


## 1 - Modèle de base

In [2]:
# Attention simple, avec self attention de base
main(train_loader, test_loader, attention = SelfAttention, modeltype = 0, lr = 1e-5)

Epoch 0 | Train loss: 0.6435007044661518 | Train accuracy: 0.6390444942381562 | Test loss: 0.6130596182276894 | Test accuracy: 0.6611715749039693
Epoch 1 | Train loss: 0.5492036817702686 | Train accuracy: 0.7231113956466069 | Test loss: 0.5247601491426537 | Test accuracy: 0.7433578745198464
Epoch 2 | Train loss: 0.5094466411750692 | Train accuracy: 0.7528809218950064 | Test loss: 0.49906211504271486 | Test accuracy: 0.7592429577464789
Epoch 3 | Train loss: 0.48859318262788914 | Train accuracy: 0.7667653649167734 | Test loss: 0.4815979176671954 | Test accuracy: 0.7703665172855314
Epoch 4 | Train loss: 0.4711230946907466 | Train accuracy: 0.7796494878361075 | Test loss: 0.465251284080263 | Test accuracy: 0.78020966709347
Epoch 5 | Train loss: 0.45599271316992335 | Train accuracy: 0.7896927016645326 | Test loss: 0.4563810602450569 | Test accuracy: 0.7847311139564661
Epoch 6 | Train loss: 0.446623133168712 | Train accuracy: 0.7941341229193342 | Test loss: 0.4414206516527107 | Test accuracy

Epoch 56 | Train loss: 0.3590773974191135 | Train accuracy: 0.839508642765685 | Test loss: 0.3660513707931498 | Test accuracy: 0.838348271446863
Epoch 57 | Train loss: 0.3586013299507231 | Train accuracy: 0.8407490396927016 | Test loss: 0.36999541847094913 | Test accuracy: 0.8360275288092189
Epoch 58 | Train loss: 0.3567524124103838 | Train accuracy: 0.8411491677336748 | Test loss: 0.3652497725763982 | Test accuracy: 0.837708066581306
Epoch 59 | Train loss: 0.3576195993473832 | Train accuracy: 0.8433098591549296 | Test loss: 0.37101287848737075 | Test accuracy: 0.8342269526248399
Epoch 60 | Train loss: 0.3565217750802846 | Train accuracy: 0.840789052496799 | Test loss: 0.36600638441884564 | Test accuracy: 0.8373079385403329
Epoch 61 | Train loss: 0.35429664280041384 | Train accuracy: 0.8421094750320103 | Test loss: 0.365229652293036 | Test accuracy: 0.8382282330345711
Epoch 62 | Train loss: 0.3552264055945504 | Train accuracy: 0.8413492317541613 | Test loss: 0.3684818820930092 | Test a

**Evolution de la loss et de l'accuracy en entraînement:**

<table>
  <tr>
    <td><img src="images/1.png" width=300 height=480></td>
    <td><img src="images/2.png" width=300 height=480></td>
  </tr>
 </table>

**Evolution de la loss et de l'accuracy en test:**

<table>
  <tr>
    <td><img src="images/3.png" width=300 height=480></td>
    <td><img src="images/4.png" width=300 height=480></td>
  </tr>
 </table>

**Observations:**

   - La loss en train converge vers la 100e époque vers 0.33
   - La loss en test converge vers la 60e époque vers 0.36
   - L'accuracy en train converge vers la 100e époque vers 0.85
   - L'accuracy en test converge vers la 60e époque vers 0.84

## 2 - Modèle résiduel

In [3]:
# Attention résiduelle, avec self attention de base
main(train_loader, test_loader, attention = SelfAttention, modeltype = 1, lr = 1e-5)

Epoch 0 | Train loss: 0.6460166187162741 | Train accuracy: 0.6393645966709347 | Test loss: 0.5881072988354442 | Test accuracy: 0.6997839308578745
Epoch 1 | Train loss: 0.5356708410531091 | Train accuracy: 0.7440781049935979 | Test loss: 0.499886618562224 | Test accuracy: 0.767485595390525
Epoch 2 | Train loss: 0.4857863849374763 | Train accuracy: 0.773927656850192 | Test loss: 0.4762629063253824 | Test accuracy: 0.7778088988476313
Epoch 3 | Train loss: 0.4670600100741191 | Train accuracy: 0.7832106274007683 | Test loss: 0.47311410310269164 | Test accuracy: 0.7783690781049936
Epoch 4 | Train loss: 0.45367445009694973 | Train accuracy: 0.7923735595390525 | Test loss: 0.44618131876201694 | Test accuracy: 0.7936939820742638
Epoch 5 | Train loss: 0.44527687325062426 | Train accuracy: 0.7968950064020487 | Test loss: 0.44503980855935665 | Test accuracy: 0.7952944942381562
Epoch 6 | Train loss: 0.43989874671538876 | Train accuracy: 0.7995758642765685 | Test loss: 0.4326145220667162 | Test accu

Epoch 56 | Train loss: 0.3629603048979702 | Train accuracy: 0.8394686299615877 | Test loss: 0.3762998160418414 | Test accuracy: 0.8317061459667093
Epoch 57 | Train loss: 0.36192250845622315 | Train accuracy: 0.8406290012804097 | Test loss: 0.37015558373082846 | Test accuracy: 0.8359875160051217
Epoch 58 | Train loss: 0.3614598657797531 | Train accuracy: 0.8405089628681178 | Test loss: 0.3672050298996527 | Test accuracy: 0.8373079385403329
Epoch 59 | Train loss: 0.36056293366135844 | Train accuracy: 0.8400688220230473 | Test loss: 0.3686222847594037 | Test accuracy: 0.8351872599231754
Epoch 60 | Train loss: 0.3593757089547021 | Train accuracy: 0.8416293213828425 | Test loss: 0.36525285068694285 | Test accuracy: 0.8373479513444302
Epoch 61 | Train loss: 0.35935390696034464 | Train accuracy: 0.8417893725992317 | Test loss: 0.3694710886289536 | Test accuracy: 0.836547695262484
Epoch 62 | Train loss: 0.3584597864582963 | Train accuracy: 0.8415092829705506 | Test loss: 0.3647002944622604 | T

**Evolution de la loss et de l'accuracy en entraînement:**

<table>
  <tr>
    <td><img src="images/5.png" width=300 height=480></td>
    <td><img src="images/6.png" width=300 height=480></td>
  </tr>
 </table>

**Evolution de la loss et de l'accuracy en test:**

<table>
  <tr>
    <td><img src="images/7.png" width=300 height=480></td>
    <td><img src="images/8.png" width=300 height=480></td>
  </tr>
 </table>

**Observations:**

   - La loss en train converge vers la 100e époque vers 0.33
   - La loss en test converge vers la 82e époque vers 0.36
   - L'accuracy en train converge vers la 100e époque vers 0.86
   - L'accuracy en test converge vers la 80e époque vers 0.84
  
Nous n'observons pas spécialement d'ajout de performance avec le modèle résiduel par rapport au modèle de base, bien qu'il était censé stabiliser le réseau.

## 3 - Ajout des plongements de position

In [4]:
# Attention résiduelle, avec self attention + Positional Encoding
main(train_loader, test_loader, attention = SelfAttentionPE, modeltype = 1, lr = 1e-5)

Epoch 0 | Train loss: 0.6916492068508066 | Train accuracy: 0.5236075544174136 | Test loss: 0.6888742719355985 | Test accuracy: 0.6019126120358514
Epoch 1 | Train loss: 0.6865827405345882 | Train accuracy: 0.5699023687580026 | Test loss: 0.6823262375005534 | Test accuracy: 0.5876680537772087
Epoch 2 | Train loss: 0.6773244434831694 | Train accuracy: 0.6017125480153649 | Test loss: 0.6695238612830715 | Test accuracy: 0.630401728553137
Epoch 3 | Train loss: 0.6621918364615202 | Train accuracy: 0.6230793854033291 | Test loss: 0.653370374525097 | Test accuracy: 0.6353233034571063
Epoch 4 | Train loss: 0.6460962670064614 | Train accuracy: 0.6363636363636364 | Test loss: 0.6412422105612736 | Test accuracy: 0.6343629961587708
Epoch 5 | Train loss: 0.6353023126053596 | Train accuracy: 0.6445262483994878 | Test loss: 0.6315162400124779 | Test accuracy: 0.6528489116517285
Epoch 6 | Train loss: 0.6256952265160642 | Train accuracy: 0.6564500640204866 | Test loss: 0.6421911486170508 | Test accuracy:

Epoch 56 | Train loss: 0.3888722093162936 | Train accuracy: 0.8261043533930857 | Test loss: 0.4004250235030387 | Test accuracy: 0.8178217029449424
Epoch 57 | Train loss: 0.38892349705729684 | Train accuracy: 0.8231033930857875 | Test loss: 0.40147714033513004 | Test accuracy: 0.8160211267605634
Epoch 58 | Train loss: 0.38807513009608935 | Train accuracy: 0.8266645326504481 | Test loss: 0.4008532447213362 | Test accuracy: 0.8167813700384123
Epoch 59 | Train loss: 0.38768919940832786 | Train accuracy: 0.8259042893725992 | Test loss: 0.39926623671331113 | Test accuracy: 0.8183418693982074
Epoch 60 | Train loss: 0.38598125102892805 | Train accuracy: 0.8280249679897568 | Test loss: 0.39916964552142253 | Test accuracy: 0.8190220870678617
Epoch 61 | Train loss: 0.3853888887859566 | Train accuracy: 0.827544814340589 | Test loss: 0.4871881054690473 | Test accuracy: 0.7710467349551856
Epoch 62 | Train loss: 0.3847005743147011 | Train accuracy: 0.8271846991037132 | Test loss: 0.39773194047785754 

**Evolution de la loss et de l'accuracy en entraînement:**

<table>
  <tr>
    <td><img src="images/9.png" width=300 height=480></td>
    <td><img src="images/10.png" width=300 height=480></td>
  </tr>
 </table>

**Evolution de la loss et de l'accuracy en test:**

<table>
  <tr>
    <td><img src="images/11.png" width=300 height=480></td>
    <td><img src="images/12.png" width=300 height=480></td>
  </tr>
 </table>

**Observations:**

   - La loss en train converge vers la 100e époque vers 0.36
   - La loss en test converge vers la 70e époque vers 0.39
   - L'accuracy en train converge vers la 100e époque vers 0.84
   - L'accuracy en test converge vers la 70e époque vers 0.82
  
Les performances en train comme en test (loss et accuracy) sont en baisse par rapport au modèle résiduel. Ce modèle a cependant l'avantage d'introduire grâce au plongement de mots une notion d'ordre sur les éléments de la séquence, ce qui l'adapte mieux aux problèmes tels que la détection de sentiment.

**Heatmap pour le positional encoding:**

<table>
  <tr>
    <td><img src="images/heatmap.png" width=450 height=500></td>
  </tr>
 </table>
 
 **Interprétation**: Le positional encoding nous permet d'ajouter une notion de séquence, d'ordre dans les données. La plus grande partie de la matrice de Positional Encoding n'est pas nécessaire afin d'encoder les dépendances entre les éléments des séquences.

## 4 - Ajout d'un token CLS

In [5]:
main(train_loader, test_loader, attention = SelfAttentionPECLS, modeltype = 1, lr = 1e-5)

Epoch 0 | Train loss: 0.6901047076366927 | Train accuracy: 0.5340508962868118 | Test loss: 0.6832238549841199 | Test accuracy: 0.584867157490397
Epoch 1 | Train loss: 0.6664255638052345 | Train accuracy: 0.6029129321382842 | Test loss: 0.6487544112756524 | Test accuracy: 0.6169174135723432
Epoch 2 | Train loss: 0.6226638505728999 | Train accuracy: 0.6559298975672215 | Test loss: 0.5959218882033828 | Test accuracy: 0.6817781690140845
Epoch 3 | Train loss: 0.5736130069030232 | Train accuracy: 0.7040653008962868 | Test loss: 0.5682133302455065 | Test accuracy: 0.7032650448143406
Epoch 4 | Train loss: 0.5432860322019966 | Train accuracy: 0.7301536491677336 | Test loss: 0.5325993090181131 | Test accuracy: 0.7331145966709347
Epoch 5 | Train loss: 0.5237674094860594 | Train accuracy: 0.7420374519846351 | Test loss: 0.5129935147095276 | Test accuracy: 0.7535211267605634
Epoch 6 | Train loss: 0.5078894564555183 | Train accuracy: 0.7535611395646606 | Test loss: 0.4992831050968048 | Test accuracy

Epoch 56 | Train loss: 0.3642544999227367 | Train accuracy: 0.8361475672215108 | Test loss: 0.3886609742019645 | Test accuracy: 0.8229433418693982
Epoch 57 | Train loss: 0.3644339251457195 | Train accuracy: 0.8376680537772087 | Test loss: 0.3905723794028354 | Test accuracy: 0.8223031370038413
Epoch 58 | Train loss: 0.36191133351068794 | Train accuracy: 0.8372679257362355 | Test loss: 0.3829298381969123 | Test accuracy: 0.8268245838668374
Epoch 59 | Train loss: 0.36129927475639784 | Train accuracy: 0.8372679257362355 | Test loss: 0.38190261145550447 | Test accuracy: 0.8289452624839949
Epoch 60 | Train loss: 0.3608925687900426 | Train accuracy: 0.8387884122919335 | Test loss: 0.3821966280891213 | Test accuracy: 0.8271046734955185
Epoch 61 | Train loss: 0.3610848000561687 | Train accuracy: 0.8378281049935979 | Test loss: 0.3879687562549103 | Test accuracy: 0.8254641485275288
Epoch 62 | Train loss: 0.35937866651084605 | Train accuracy: 0.8381081946222791 | Test loss: 0.3812653160571728 | T

**Evolution de la loss et de l'accuracy en entraînement:**

<table>
  <tr>
    <td><img src="images/13.png" width=300 height=480></td>
    <td><img src="images/14.png" width=300 height=480></td>
  </tr>
 </table>

**Evolution de la loss et de l'accuracy en test:**

<table>
  <tr>
    <td><img src="images/15.png" width=300 height=480></td>
    <td><img src="images/16.png" width=300 height=480></td>
  </tr>
 </table>

**Observations:**

   - La loss en train converge vers la 100e époque vers 0.32
   - La loss en test converge vers la 60e époque vers 0.38
   - L'accuracy en train converge vers la 100e époque vers 0.86
   - L'accuracy en test converge vers la 60e époque vers 0.83
  
Par rapport au modèle résiduel avec positional encoding mais sans token CLS, celui-là donne de meilleures performances de manières générale: meilleures losses et accuracies en train comme en test, convergence plus rapide en test également. Cela montre l'intérêt du token CLS afin de représenter le sens de toute la séquence (représentation de toute la phrase) pour la classification.